### DISCUSSION TITLE
Bounty proposed: Restart sleep timer via play button on headphones (in addition to shaking)

### DISCUSSION THREAD

**Discussion by bjohas** [PaulWoitaschek/Voice#1288] [Ideas]
It's amazing that 
* Voice has got the shake feature, to restart the sleep timer.
* Also, it's great that the restart now only happens within 30 secs of the sleep timer expiring. I.e., it's no longer possible for the book to be accidentally restarted when the device is picked up later (In Version 5.0.2, https://github.com/PaulWoitaschek/Voice/issues/954).

## Feature request

I would like to be able to restart the timer via a play/pause button, typically on headphones (wired or bluetooth). For example, when listening on headphones or via a pillow-speaker, the phone may not be all that close to allow shaking to restart; however, it may be easy to restart the player via the play/pause button on headphones.

The issues https://github.com/PaulWoitaschek/Voice/issues/742, https://github.com/PaulWoitaschek/Voice/issues/988 are related to this. There are two options: 

1. The play/pause button has to be pressed within the 30 seconds after sleep-timer expiry.
2. The sleep timer persists, i.e., play/pause button can be pressed any time after sleep-timer expiry to restart the sleep timer. However, the shake timer would be unaffected (i.e., expire 30 seconds after sleep-timer expiry).

Edit (2022-02-25): I can see that Option 1 is consistent with the shake behaviour. <s>However, I can see Option 2 being more useful: e.g., people forgetting to put the sleep timer on (https://github.com/PaulWoitaschek/Voice/issues/742) or simply restarting listening again without having to set up the sleep timer. Therefore, I prefer Option 2.</s> Option 1 may have a simple fix (see below) and does meet my use case - so I'd be perfectly happy with it.

## UI

There would probably need to be an option in the UI somewhere. I don't really mind how that's done. Obviously, Voice takes a lean approach, so the option would need to be unobtrusive. A simple option, proposed in the other issues, is to add a check-box to the sleep timer menu, e.g. 'resume sleep timer on play' or similar.

## Bounty

If @PaulWoitaschek agrees with the change (and would accept a PR), I'd like to offer a USD 50 bounty for the implementation of this feature (implemented as a PR on https://github.com/PaulWoitaschek/Voice as well as a downloadable apk, based off 5.0.2). (@PaulWoitaschek - if you're interested, I'd obviously also be happy to pay the bounty to you.)

Any interest?

**Comment by bjohas** [comment]
Hi @PaulWoitaschek - any thoughts on this? Would a PR be welcome?

**Comment by nydrani** [comment]
I have a general idea on how this can be completed, the sleeptimer needs to receive some kind of signal from the service (MediaSessionCompat.Callback) triggered, either from a direct callback or via other channels e.g. a global singleton source OR via a broadcast source). It is difficult to directly code a callback as the player service is decoupled from the sleeptimer.

**Comment by bjohas** [comment]
Hi @nydrani 

I've added a simple idea via the PR $1154. Basically, I have **not** compiled this (as I don't have those skills), but conceptually, the idea is
* Let sleep timer exit.
* If the player is playing at that stage, restart the sleep timer.

This has the disadvantage that the new period is essentially the sleep period plus the timeout period, but to me that really doesn't matter.

The idea is this. Here https://github.com/PaulWoitaschek/Voice/blob/main/sleepTimer/src/main/kotlin/voice/sleepTimer/SleepTimer.kt consider `start()`.

```
  private fun start() {
    Logger.i("Starting sleepTimer. Pause in $sleepTime.")
    leftSleepTime = sleepTime
    sleepJob?.cancel()
    sleepJob = scope.launch {
      startSleepTimerCountdown()
      val shakeToResetTime = 30.seconds
      Logger.d("Wait for $shakeToResetTime for a shake")
      withTimeout(shakeToResetTime) {
        shakeDetector.detect()
        Logger.i("Shake detected. Reset sleep time")
        playerController.play()
        start()
      }
     // HERE
      Logger.i("exiting")
    }
  }
```

At the point marked `HERE`, the sleep timer times out. However, even while the sleep timer is timing out, I can actually press play. So what if the play state is playing at the moment the sleep timer times out? In code, how about this?

```
  private fun start() {
    Logger.i("Starting sleepTimer. Pause in $sleepTime.")
    leftSleepTime = sleepTime
    sleepJob?.cancel()
    sleepJob = scope.launch {
      startSleepTimerCountdown()
      val shakeToResetTime = 30.seconds
      Logger.d("Wait for $shakeToResetTime for a shake")
      withTimeout(shakeToResetTime) {
        shakeDetector.detect()
        Logger.i("Shake detected. Reset sleep time")
        playerController.play()
        start()
      }
     // HERE
      if (playStateManager.playState == Playing) {
        Logger.i("Sleep timer finished, player is playing, restart timer.")
        start()
      }
      Logger.i("exiting")
    }
  }
```

The sleep timer exists, and if the play state is playing, then restart the sleep timer.

**Reply by bjohas** [reply]
Hi @PaulWoitaschek - thanks for converting this into a discussion. I'd still be very keen on this feature!

**Comment by SantoshGuptaML** [comment]
A hacky way to do this 

```
  fun playPause() {
    player.playPause()
    sleepTimer.setActive(true)
  }

```

https://github.com/PaulWoitaschek/Voice/blob/000f7f94b711686b70082f9ebcf26f12fe7009a5/playbackScreen/src/main/kotlin/voice/playbackScreen/BookPlayViewModel.kt#L75
